Create a macro to generate BitArray variations

taikiy opened the pull request:
Similar to `Field`, create a macro to generate BitArray structs with a specified bit length. Added `BitArray40` for the match key.

`BitArray64` is intact which we could use for other report fields i.e. breakdown key, but I suppose we'll use a lot shorter bit-length struct for that.

taikiy commented on src/bits/bit_array.rs:
```diff
@@ -1,193 +1,199 @@
+        }
+
+        impl std::ops::Index<usize> for $array {
+            type Output = u8;
```
Thanks! Removed the custom `Index` and `TryFrom`, and all other trait implementations that are provided by a bare BitArray.

martinthomson commented on src/bits/mod.rs:
```diff
@@ -1,19 +1,16 @@
-    /// into this data type. This conversion is lossy. Callers are encouraged
-    /// to use `try_from` if the input is not known in advance.
+    /// into this data type. This conversion is lossy.
     fn truncate_from<T: Into<u128>>(v: T) -> Self;
```
This could be marked `#[cfg(test)]` to avoid accidental misuse.

taikiy replied:
This will be used in production. I spoke with Alex and this is the name we agreed on for the lossy conversion. Alternatively, we could use BitSlice or bitarr! macro to create a new BiArray instance, but neither is not user-friendly. I think we'll almost always use this function. Or we might need to bring back TryFrom<u128>.

martinthomson reviewed (commented):
WFM

taikiy commented:
Per discussion offline with @martinthomson. Addressed PR comments other than removing `TryFrom<u128>` and using a bare BitArray. I think we'll need a wrapper struct to implement i.e. `Distribution` for the bit array structs for `IntoShares`.

benjaminsavage reviewed (commented):
Thanks for adding this @taikiy
